// @sentry/node/dist/parsers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = require("@sentry/utils/fs");
var path_1 = require("@sentry/utils/path");
var string_1 = require("@sentry/utils/string");
var stacktrace = require("stack-trace");
var LINES_OF_CONTEXT = 7;
/** JSDoc */
function getFunction(frame) {
    try {
        return frame.getFunctionName() || frame.getTypeName() + "." + (frame.getMethodName() || '<anonymous>');
    }
    catch (e) {
        // This seems to happen sometimes when using 'use strict',
        // stemming from `getTypeName`.
        // [TypeError: Cannot read property 'constructor' of undefined]
        return '<anonymous>';
    }
}
var mainModule = ((require.main && require.main.filename && path_1.dirname(require.main.filename)) ||
    global.process.cwd()) + "/";
/** JSDoc */
function getModule(filename, base) {
    if (!base) {
        base = mainModule; // tslint:disable-line:no-parameter-reassignment
    }
    // It's specifically a module
    var file = path_1.basename(filename, '.js');
    filename = path_1.dirname(filename); // tslint:disable-line:no-parameter-reassignment
    var n = filename.lastIndexOf('/node_modules/');
    if (n > -1) {
        // /node_modules/ is 14 chars
        return filename.substr(n + 14).replace(/\//g, '.') + ":" + file;
    }
    // Let's see if it's a part of the main module
    // To be a part of main module, it has to share the same base
    n = (filename + "/").lastIndexOf(base, 0);
    if (n === 0) {
        var moduleName = filename.substr(base.length).replace(/\//g, '.');
        if (moduleName) {
            moduleName += ':';
        }
        moduleName += file;
        return moduleName;
    }
    return file;
}
/** JSDoc */
function readSourceFiles(filenames) {
    return __awaiter(this, void 0, void 0, function () {
        var sourceFiles;
        var _this = this;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    // we're relying on filenames being de-duped already
                    if (filenames.length === 0) {
                        return [2 /*return*/, {}];
                    }
                    sourceFiles = {};
                    return [4 /*yield*/, Promise.all(filenames.map(function (filename) { return __awaiter(_this, void 0, void 0, function () {
                            var content, _1;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        _a.trys.push([0, 2, , 3]);
                                        return [4 /*yield*/, fs_1.readFileAsync(filename)];
                                    case 1:
                                        content = _a.sent();
                                        return [3 /*break*/, 3];
                                    case 2:
                                        _1 = _a.sent();
                                        // unsure what to add here as the file is unreadable
                                        content = null;
                                        return [3 /*break*/, 3];
                                    case 3:
                                        if (typeof content === 'string') {
                                            sourceFiles[filename] = content;
                                        }
                                        return [2 /*return*/];
                                }
                            });
                        }); }))];
                case 1:
                    _a.sent();
                    return [2 /*return*/, sourceFiles];
            }
        });
    });
}
/** JSDoc */
function extractStackFromError(error) {
    return __awaiter(this, void 0, void 0, function () {
        var stack;
        return __generator(this, function (_a) {
            stack = stacktrace.parse(error);
            if (!stack) {
                return [2 /*return*/, []];
            }
            return [2 /*return*/, stack];
        });
    });
}
exports.extractStackFromError = extractStackFromError;
/** JSDoc */
function parseStack(stack) {
    return __awaiter(this, void 0, void 0, function () {
        var filesToRead, frames, _2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    filesToRead = [];
                    frames = stack.map(function (frame) {
                        var parsedFrame = {
                            colno: frame.getColumnNumber(),
                            filename: frame.getFileName() || '',
                            function: getFunction(frame),
                            lineno: frame.getLineNumber(),
                        };
                        var isInternal = frame.isNative() ||
                            (parsedFrame.filename &&
                                !parsedFrame.filename.startsWith('/') &&
                                !parsedFrame.filename.startsWith('.') &&
                                parsedFrame.filename.indexOf(':\\') !== 1);
                        // in_app is all that's not an internal Node function or a module within node_modules
                        // note that isNative appears to return true even for node core libraries
                        // see https://github.com/getsentry/raven-node/issues/176
                        parsedFrame.in_app =
                            !isInternal && parsedFrame.filename !== undefined && !parsedFrame.filename.includes('node_modules/');
                        // Extract a module name based on the filename
                        if (parsedFrame.filename) {
                            parsedFrame.module = getModule(parsedFrame.filename);
                            if (!isInternal) {
                                filesToRead.push(parsedFrame.filename);
                            }
                        }
                        return parsedFrame;
                    });
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, addPrePostContext(filesToRead, frames)];
                case 2: return [2 /*return*/, _a.sent()];
                case 3:
                    _2 = _a.sent();
                    // This happens in electron for example where we are not able to read files from asar.
                    // So it's fine, we recover be just returning all frames without pre/post context.
                    return [2 /*return*/, frames];
                case 4: return [2 /*return*/];
            }
        });
    });
}
exports.parseStack = parseStack;
/**
 * This function tries to read the source files + adding pre and post context (source code)
 * to a frame.
 * @param filesToRead string[] of filepaths
 * @param frames StackFrame[] containg all frames
 */
function addPrePostContext(filesToRead, frames) {
    return __awaiter(this, void 0, void 0, function () {
        var sourceFiles;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, readSourceFiles(filesToRead)];
                case 1:
                    sourceFiles = _a.sent();
                    return [2 /*return*/, frames.map(function (frame) {
                            if (frame.filename && sourceFiles[frame.filename]) {
                                try {
                                    var lines = sourceFiles[frame.filename].split('\n');
                                    frame.pre_context = lines
                                        .slice(Math.max(0, (frame.lineno || 0) - (LINES_OF_CONTEXT + 1)), (frame.lineno || 0) - 1)
                                        .map(function (line) { return string_1.snipLine(line, 0); });
                                    frame.context_line = string_1.snipLine(lines[(frame.lineno || 0) - 1], frame.colno || 0);
                                    frame.post_context = lines
                                        .slice(frame.lineno || 0, (frame.lineno || 0) + LINES_OF_CONTEXT)
                                        .map(function (line) { return string_1.snipLine(line, 0); });
                                }
                                catch (e) {
                                    // anomaly, being defensive in case
                                    // unlikely to ever happen in practice but can definitely happen in theory
                                }
                            }
                            return frame;
                        })];
            }
        });
    });
}
/** JSDoc */
function getExceptionFromError(error) {
    return __awaiter(this, void 0, void 0, function () {
        var name, stack, frames;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    name = error.name || error.constructor.name;
                    return [4 /*yield*/, extractStackFromError(error)];
                case 1:
                    stack = _a.sent();
                    return [4 /*yield*/, parseStack(stack)];
                case 2:
                    frames = _a.sent();
                    return [2 /*return*/, {
                            stacktrace: {
                                frames: prepareFramesForEvent(frames),
                            },
                            type: name,
                            value: error.message,
                        }];
            }
        });
    });
}
exports.getExceptionFromError = getExceptionFromError;
/** JSDoc */
function parseError(error) {
    return __awaiter(this, void 0, void 0, function () {
        var e_1, _a, _b, name, exception, event, errorKeys, extraErrorInfo, errorKeys_1, errorKeys_1_1, key;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    name = error.name || error.constructor.name;
                    return [4 /*yield*/, getExceptionFromError(error)];
                case 1:
                    exception = _c.sent();
                    event = {
                        exception: {
                            values: [exception],
                        },
                        message: name + ": " + (error.message || '<no message>'),
                    };
                    errorKeys = Object.keys(error).filter(function (key) { return !(key in ['name', 'message', 'stack', 'domain']); });
                    if (errorKeys.length) {
                        extraErrorInfo = {};
                        try {
                            for (errorKeys_1 = __values(errorKeys), errorKeys_1_1 = errorKeys_1.next(); !errorKeys_1_1.done; errorKeys_1_1 = errorKeys_1.next()) {
                                key = errorKeys_1_1.value;
                                extraErrorInfo[key] = error[key];
                            }
                        }
                        catch (e_1_1) { e_1 = { error: e_1_1 }; }
                        finally {
                            try {
                                if (errorKeys_1_1 && !errorKeys_1_1.done && (_a = errorKeys_1.return)) _a.call(errorKeys_1);
                            }
                            finally { if (e_1) throw e_1.error; }
                        }
                        event.extra = (_b = {},
                            _b[name] = extraErrorInfo,
                            _b);
                    }
                    return [2 /*return*/, event];
            }
        });
    });
}
exports.parseError = parseError;
/** JSDoc */
function prepareFramesForEvent(stack) {
    if (!stack || !stack.length) {
        return [];
    }
    var localStack = stack;
    var firstFrameFunction = localStack[0].function || '';
    if (firstFrameFunction.includes('captureMessage') || firstFrameFunction.includes('captureException')) {
        localStack = localStack.slice(1);
    }
    // The frame where the crash happened, should be the last entry in the array
    return localStack.reverse();
}
exports.prepareFramesForEvent = prepareFramesForEvent;
